File: BEMLIB_22.10/01_num_meth/04_nl_eq/quad.f
      program quad

c=========================================
c FDLIB
c
c Copyright by C. Pozrikidis, 1999
c All rights reserved
c
c This program is to be used only under the
c stipulations of the licensing agreement
c=========================================

c------------------------------------------------
c This program accompanies the book:
c
c            C. Pozrikidis
c Numerical Computation in Science and Engineering
c       Oxford University Press
c------------------------------------------------

c------------------------------
c roots of a quadratic equation
c with real coefficients a, b, c
c------------------------------

      Implicit Double Precision (a-h,o-z)

      write (6,*)
      write (6,*) " Roots of: a*x^2 + b*x + c = 0"

 98   Continue

      write (6,*)
      write (6,*) "Please enter the coefficients a, b, c"
      write (6,*) "99 for any one to quit"
      write (6,*) "----------------------"
      read  (5,*) a,b,c 

      if(a.eq.99) Go to 99
      if(b.eq.99) Go to 99
      if(c.eq.99) Go to 99

c------
c traps
c------

      if(abs(a).lt.0.0000001) then

         if(abs(b).gt.0.0000001) then
           write (6,*) 
           write (6,*) " The coefficient a is too small"
           write (6,*) " and the equation looks linear"
           root = -c/b
           write (6,*) " The root is: ", root
         else
           write (6,*) " No root for the constant quadratic"
         end if

         Go to 98

      end if

c-------------------------------------
c proceed with the analytical solution
c-------------------------------------

      d  = b*b-4.0D0*a*c   ! discriminant
      a2 = 2.0D0*a

      if(d.ge.0) then

         srd = Dsqrt(d)
         x1=(-b+srd)/a2
         x2=(-b-srd)/a2

         write (6,*)
         write (6,*) 'roots: ',x1,x2

      else

         d = -d
         real = -b/a2
         rmag = Dsqrt(d)/a2

         write (6,*)
         write (6,*) 'Roots are complex conjugate'
         write (6,*)
         write (6,*) 'Real part:      ',real
         write (6,*) 'Imaginary part: ',rmag

      end if

      Go to 98

c-----
c Done
c-----

  99  Continue

      stop
      end 

File: BEMLIB_22.10/01_num_meth/04_nl_eq/quad.cc
/*------------------------------
  roots of a quadratic equation
 ------------------------------*/

#include<iostream>
#include<cmath>

using namespace std;

int main()
{
   double a,b,c,d,a2,srd,x1,x2,imag,real,root;
   cout<<"\nRoots of: a*x**2 + b*x + c = 0\n";

   while(1)
   {
      cout<<"\n Please enter the coefficients a, b, c";
      cout<<"\n 999 for any one to quit";
      cout<<"\n------------------------\n";
      cin>>a;
      cin>>b;
      cin>>c;
      if( (a==999) || (b==999) || (c==999) )
      break;

      /*------
      c traps
      c------*/

      if(abs(a) < 0.0000001)
        {
         cout<<"\nThe coefficient a is too small";
         cout<<"\nand the equation looks linear\n";

           if(abs(b) > 0.0000001)
            {
             root = -c/b;
             cout<<"\nThe root is: "<<root<<"\n";
            }
        else
          cout<<"\nNo root for the constant binomial\n";
        }

/*-------------------
  analytical solution
 --------------------*/

     else
     {
       d = b*b-4.0*a*c;   //discriminant
       a2 = 2.0*a;
       if(d>=0)
        {
         srd = sqrt(d);
         x1=(-b+srd)/a2;
         x2=(-b-srd)/a2;
         cout<<"\nroots are:"<<x1<<", "<<x2<<"\n";
        }
         else
        {
          d = -d;
          imag = sqrt(d)/a2;
          real = -b/a2;
          cout<<"\n roots are complex conjugate";
          cout<<"\n real part:\t"<<real;
          cout<<"\n imaginary part:\t"<<imag<<"\n";
        }
       }

    } //end while-loop

 return 0;
 }
